### PR TITLE
add TSX support

### LINKED ISSUE #2543: Support TSX examples
Support TSX in Example.js to allow editing TS
  - Have to transpile TSX to JSX in-browser
  - Currently using Acorn to transpile JSX to JS
    - Acorn can’t transpile typescript? https://github.com/g-plane/acorn-typescript/blob/master/index.js
    - Need to find/write typescript acorn plugin
    
How can we recruit contributors to add TSX examples? Should we add a shiny label to the "TSX" button, or add a grayed out "TSX" button with a tooltip to all react-core examples?

image

### PR DESCRIPTION
**Pull request by redallen**
Closes #2543

Typescript test file: https://github.com/patternfly/patternfly-react/blob/main/packages/react-integration/demo-app-ts/src/components/demos/AboutModal/AboutModalDemo.tsx

Replaces react-live:
  - Uses our own `@patternfly/react-code-editor` as the editor
    - Uses its toolbar as ExampleToolbar
      - Needed some tweaks to hide editor, so bump `@patternfly/react-docs`
  - Use `new Function` statement to eval live code
  - Use `componentDidCatch` in new `class ErrorBoundary` to catch component errors

Adds TSX support:
  - Explicitly support ES2018 in live examples
  - Add `acorn-typescript.js` to parse types
    - Forked from https://github.com/g-plane/acorn-typescript/blob/master/index.js (MIT licensed)
  - Detect language live in-parser by presence of types
  - Add button to convert TS to JS using modified `astring` generator
  - Centralize `liveCodeTypes` for easier future language modifications

Quality of life:
- Add "undo all" button
- Add ```tsx file="filename" support in MD
- Debounce transpiling live code
- Error message appears in example region above editor (where the user is likely looking) instead of below editor
- Editor has a max height

Future work:
- Enable typechecking after https://github.com/patternfly/patternfly-react/pull/6035 is merged. Will need a way to concatenate type files in `@patternfly` packages for patternfly types to work.

### CODE REVIEW

**Review by jschuler — COMMENTED** [review]
Looks and works great, but adding some comments and having some test code would be very beneficial!

File: packages/theme-patternfly-org/components/example/example.js [hunk 1]
```diff
@@ -1,35 +1,60 @@
...
+      return errorComponent(this.state.error);
+    }
+    return this.props.children;
+  }  
 }
 
 // Props come from mdx-ast-to-mdx-hast.js
 export const Example = ({
   code,
   lang = '',
```
[the remark this answers is not in the document]
Comment (redallen) [reply packages/theme-patternfly-org/components/example/example.js:57]: Done!

File: packages/theme-patternfly-org/components/example/exampleToolbar.js [hunk 2]
```diff
@@ -1,122 +1,169 @@
...
-    openLang: null,
-    showCopyMessage: false,
-  };
+  return lang;
+}
 
-  onCopy = () => {
-    copy(this.props.code);
+export const ExampleToolbar = ({
+  fullscreenLink,
```
[the remark this answers is not in the document]
Comment (redallen) [reply packages/theme-patternfly-org/components/example/exampleToolbar.js:24]: Done!

**Comment by jschuler** [comment]
LGTM, do you think it would be helpful to add some tests for the transformation? If so we can open a separate issue

**Review by jschuler — APPROVED** [review]

**Comment by redallen** [comment]
It would be helpful to open an issue for, yes.

Currently there's no testing framework in org so I'd have to pick the best one and set it up.

**Pull request merged**